In [84]:
import numpy as np
import pandas as pd

In [2]:
car = pd.read_csv('RAV4-142-Spring2021.csv')
car

MonthNumeric MonthFactor  Year  RAV4Sales  Unemployment  RAV4Queries  \
0               1     January  2011      11196           9.1           29   
1               2    February  2011      12562           9.0           29   
2               3       March  2011      16082           9.0           29   
3               4       April  2011      15586           9.1           27   
4               5         May  2011       8624           9.0           28   
..            ...         ...   ...        ...           ...          ...   
115             8      August  2020      39239           8.4           94   
116             9   September  2020      43652           7.8           88   
117            10     October  2020      40717           6.9           89   
118            11    November  2020      40250           6.7           79   
119            12    December  2020      46846           6.7           85   

      CPIAll  CPIEnergy  
0    221.187    229.258  
1    221.898    232.068  
2    223.046    240.079  
3    224.093    247.977  
4    224.806    250.744  
..       ...        ...  
115  259.681    194.430  
116  260.209    195.990  
117  260.325    196.269  
118  260.817    197.112  
119  261.775    205.084  

[120 rows x 8 columns]

## Part A

In [13]:
car_train = car[car['Year'] <= 2016]
car_test = car[car['Year'] >= 2017]
car_test;

## Training the Model

In [4]:
import statsmodels.formula.api as smf

ols = smf.ols(formula='RAV4Sales ~ Unemployment + RAV4Queries + CPIAll + CPIEnergy', data=car_train)
model1 =ols.fit()
print(model1.summary())

                            OLS Regression Results                            
Dep. Variable:              RAV4Sales   R-squared:                       0.810
Model:                            OLS   Adj. R-squared:                  0.799
Method:                 Least Squares   F-statistic:                     71.42
Date:                Tue, 09 Feb 2021   Prob (F-statistic):           1.93e-23
Time:                        15:55:33   Log-Likelihood:                -683.31
No. Observations:                  72   AIC:                             1377.
Df Residuals:                      67   BIC:                             1388.
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
Intercept     1.961e+04   9.36e+04      0.210   

In [5]:
# MATRIX STYLE TRAINING 
# Import the library that contains all the functions/modules related to the regression model
import statsmodels.api as sm

# Choose the features to be used
cols = ['Unemployment', 'RAV4Queries', 'CPIAll', 'CPIEnergy']
X_train = car_train[cols]
y_train = car_train['RAV4Sales']

# We must add an intercept as the standard model doesn't automatically fit one
X_train = sm.add_constant(X_train)

# Fit the data to the model
model1 = sm.OLS(y_train, X_train).fit()
print(model1.summary())

                            OLS Regression Results                            
Dep. Variable:              RAV4Sales   R-squared:                       0.810
Model:                            OLS   Adj. R-squared:                  0.799
Method:                 Least Squares   F-statistic:                     71.42
Date:                Tue, 09 Feb 2021   Prob (F-statistic):           1.93e-23
Time:                        15:55:33   Log-Likelihood:                -683.31
No. Observations:                  72   AIC:                             1377.
Df Residuals:                      67   BIC:                             1388.
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
const         1.961e+04   9.36e+04      0.210   

In [6]:
# calculate Variance Inflation Factor for each explanatory variable
from statsmodels.stats.outliers_influence import variance_inflation_factor

# The dataframe passed to VIF must include the intercept term. We add it the same way we did before.
def VIF(df, columns):
    values = sm.add_constant(df[columns]).values
    num_columns = len(columns)+1
    vif = [variance_inflation_factor(values, i) for i in range(num_columns)]
    return pd.Series(vif[1:], index=columns)

cols = ['Unemployment', 'RAV4Queries', 'CPIAll', 'CPIEnergy']
VIF(car_train, cols)

Unemployment    37.437684
RAV4Queries      6.231404
CPIAll          28.216088
CPIEnergy        7.220536
dtype: float64

In [7]:
# remove unemployment
model3 = smf.ols(formula='RAV4Sales ~ RAV4Queries + CPIAll + CPIEnergy', data=car_train).fit()
print(model3.summary())

cols = ['RAV4Queries', 'CPIAll', 'CPIEnergy']
VIF(car_train, cols)

                            OLS Regression Results                            
Dep. Variable:              RAV4Sales   R-squared:                       0.801
Model:                            OLS   Adj. R-squared:                  0.792
Method:                 Least Squares   F-statistic:                     91.21
Date:                Tue, 09 Feb 2021   Prob (F-statistic):           8.73e-24
Time:                        15:56:38   Log-Likelihood:                -684.99
No. Observations:                  72   AIC:                             1378.
Df Residuals:                      68   BIC:                             1387.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
Intercept   -1.379e+05   3.21e+04     -4.290      

RAV4Queries    5.952056
CPIAll         3.876882
CPIEnergy      2.610675
dtype: float64

In [ ]:
# remove RAV4Queries
model2 = smf.ols(formula='RAV4Sales ~  CPIAll + CPIEnergy', data=car_train).fit()
print(model2.summary())

cols = ['CPIAll', 'CPIEnergy']
VIF(car_train, cols)

In [10]:
# remove CPIAll from original
model2 = smf.ols(formula='RAV4Sales ~ Unemployment + RAV4Queries + CPIEnergy', data=car_train).fit()
print(model2.summary())

cols = ['Unemployment', 'RAV4Queries', 'CPIEnergy']
VIF(car_train, cols)

                            OLS Regression Results                            
Dep. Variable:              RAV4Sales   R-squared:                       0.810
Model:                            OLS   Adj. R-squared:                  0.802
Method:                 Least Squares   F-statistic:                     96.62
Date:                Tue, 09 Feb 2021   Prob (F-statistic):           1.81e-24
Time:                        16:09:13   Log-Likelihood:                -683.32
No. Observations:                  72   AIC:                             1375.
Df Residuals:                      68   BIC:                             1384.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
Intercept     3.093e+04   9228.813      3.351   

Unemployment    5.143926
RAV4Queries     4.857967
CPIEnergy       2.965016
dtype: float64

In [11]:
# remove CPIEnergy
model2 = smf.ols(formula='RAV4Sales ~ Unemployment + RAV4Queries', data=car_train).fit()
print(model2.summary())

cols = ['Unemployment', 'RAV4Queries']
VIF(car_train, cols)

                            OLS Regression Results                            
Dep. Variable:              RAV4Sales   R-squared:                       0.810
Model:                            OLS   Adj. R-squared:                  0.804
Method:                 Least Squares   F-statistic:                     147.1
Date:                Tue, 09 Feb 2021   Prob (F-statistic):           1.31e-25
Time:                        16:10:40   Log-Likelihood:                -683.32
No. Observations:                  72   AIC:                             1373.
Df Residuals:                      69   BIC:                             1379.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
Intercept     3.118e+04   7016.566      4.443   

Unemployment    4.469452
RAV4Queries     4.469452
dtype: float64

## Understand the Data
i) The linear regression equation produced by the model is RAV4Sales = -3046.6355 * Unemployment + 245.9654 * RAV4Queries. This means for every 1 unit increase in unemployment, RAV4Sales will decrease by 3046.6355; for every 1 unit increase in RAV4Queries, RAV4Sales will increase by 245.9654.

ii) I selected my variables by first performing ordinary least squares (OLS) on the model with all 4 independent variables. I noticed that Unemployment and CPIAll had very high VIFs scores, but CPIAll had a significantly higher p-value than Unemployment. Thus I removed CPIAll and ran the analysis again. This made VIF scores for the remaining 3 independent variables less than 6, so I looked at the p-values. CPIEnergy had a p-value of 0.967, which is very high, so I removed it. Then the remaining Unemployment and RAV4Queries had VIF values less than 5, and very close to 0, which is excellent. The adjusted R^2 for this model was 0.804, which is better than the original model's adjusted R^2 of .799.

iii) The signs of the coefficients make sense, because as unemployment increases, people have less money and won't spend as much. Thus Unemployment and RAV4Sales have an inverse relationship. It makes sense that RAV4Queries has a positive coefficient because as more people search for RAV4s, this indicates a higher demand and more people buying it.

iv) The model is a good predictor of training set observations, as the R-square is 0.810. This is semi-close to 1, which we want.


### Part B

In [16]:
import statsmodels.formula.api as smf

ols = smf.ols(formula='RAV4Sales ~ MonthFactor + Unemployment + RAV4Queries + CPIAll + CPIEnergy', data=car_train)
model5 =ols.fit()
print(model5.summary())

# cols = ['Unemployment', 'RAV4Queries', 'CPIAll', 'CPIEnergy']
# VIF(car_train, cols)

                            OLS Regression Results                            
Dep. Variable:              RAV4Sales   R-squared:                       0.884
Model:                            OLS   Adj. R-squared:                  0.853
Method:                 Least Squares   F-statistic:                     28.51
Date:                Tue, 09 Feb 2021   Prob (F-statistic):           8.55e-21
Time:                        16:25:28   Log-Likelihood:                -665.48
No. Observations:                  72   AIC:                             1363.
Df Residuals:                      56   BIC:                             1399.
Df Model:                          15                                         
Covariance Type:            nonrobust                                         
                               coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------------
Intercept               

Unemployment    37.437684
RAV4Queries      6.231404
CPIAll          28.216088
CPIEnergy        7.220536
dtype: float64

i) The regression equation is RAV4Sales = 7.754e+04 + 2422.0989*August + 1885.2249*December + -2922.8349*February + -4543.8071*January + -193.7079*July + -1426.1733*June + 466.8540*March + 2010.0694*May + -1540.1770*November + -1808.4239*October + 1879.2594*September + -3687.3648*Unemployment + 228.4423*RAV4Queries + -175.4430*CPIAll + 1.1895*CPIEnergy

The  coefficients  of  each  of  the $MonthFactor$ dummy variables represent that marginal impact that the specific month in the year has on the sales of the car.

ii) The training set $R^2$ for this model is 0.884. The significant variables are Unemployment and RAV4Queries, and January, as they have the lowest p-values. The months with the greatest impact are January and February; these are the only months with p-value less than 0.05. 

iii) Yes, I think adding the independent variable $MonthFactor$ improves the model, because the $R^2$ increased by 0.074 from the previous model. Furthermore, since the p-value is less than 0.05 for January (and close to 0.05 for February), the result of the hypothesis test suggests that it is important to consider MonthFactor and hence keep it.

iv) Since the months with the greatest impact are 2 consecutive months, another way to model seasonality could be to categorize by season, ie Winter/Spring/Summer/Fall. This would decrease the number of dummy variables and perhaps the VIF too since February sales should be similar to January sales. An additional method is to create dummy variables based on the year.

### Part C
Build a final model using a subset of the independent variables used in parts (a) and (b), providing a brief justification for the variables selected.  What is the training setR2and theOSR2(this is theR2of your model on the test set)?  Do you think your model wouldbe useful to Toyota?  Why or why not?

In [30]:
import statsmodels.formula.api as smf

ols = smf.ols(formula= 'RAV4Sales ~ MonthFactor + Unemployment + RAV4Queries', data=car_test) #change to car_train
modelc =ols.fit()
print(modelc.summary())

# cols = ['Unemployment', 'RAV4Queries', 'CPIAll', 'CPIEnergy']
# VIF(car_train, cols)

                            OLS Regression Results                            
Dep. Variable:              RAV4Sales   R-squared:                       0.721
Model:                            OLS   Adj. R-squared:                  0.614
Method:                 Least Squares   F-statistic:                     6.761
Date:                Tue, 09 Feb 2021   Prob (F-statistic):           3.67e-06
Time:                        17:16:17   Log-Likelihood:                -460.43
No. Observations:                  48   AIC:                             948.9
Df Residuals:                      34   BIC:                             975.1
Df Model:                          13                                         
Covariance Type:            nonrobust                                         
                               coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------------
Intercept               

My final model includes as independent variables: MonthFactor, Unemployment, RAV4Queries. I chose these because Unemployment amd RAV4Queries were the two that I decided to keep from part a, and in part b, I decided to keep MonthFactor. The training set R-square is 0.883, and the OSR-square is 0.721. I think that my model could be useful to Toyota because the OSR-square is pretty good, but not great. My model would be more useful if I had data that was more specific to this problem, like the number of visits to Toyota's RAV4 sales page or the number of pre-orders.

## Part D: Aggregating my Own Data from Another Source
I chose to add data based on monthly car loan rates. My hypothesis is people are more likely or willing to buy cars when loan rates are low. I know from Economics that the Federal Reserve decreases interest rates when it wants to encourage people to spend and stimulate the economy, thus I predicted it could indicate spending on cars. I obtained my data from Statista (https://www.statista.com/statistics/290673/auto-loan-rates-usa/): Interest rates on auto loans in the United States from January 2014 to December 2020. 

Then I spent a long time cleaning and preparing the interest rate data before merging it with the original Toyota car sales data. I used the same independent variables as in part (c), with the addition of LoanRate. My R-square for this new model is 0.844, sligltly worse than the R-square from part (c). Additionally, the p-value for LoanRate in the training set is 0.751, which is extremely high. This suggests that it didn't add any predictive value. One thing to note however, is that the data I had only started in 2014, while the Toyota data started in 2011, so if I had access to the previous years, I could've improved the data. Furthermore, the coefficient for LoanRate is -1517.0760, which makes sense that it's negative because people buy less cars as the loan rate increases.

Testing my model on the test set, by changing the ols code to 'data=data_test' gives an OSR-square of 0.773. This is slightly better than the OSR-square from part(c), which suggesets that having the LoanRate data for all the years in the test set improves the model. Therefore I believe that the LoanRate data adds predictive value.

In [114]:
rate = pd.read_excel('statistic_id290673_monthly-car-loan-rates-in-the-us-2014-2020.xlsx', sheet_name ='Data')
rate['MonthNumeric'] = pd.to_datetime(rate.MonthFactor, format='%b').dt.month
rate = rate.drop(columns = ['MonthFactor'])
data = pd.merge(car, rate, how='left', on=['Year', 'MonthNumeric'])
data_train = data[data['Year'] <= 2016]
data_test = data[data['Year'] >= 2017]
data_train

MonthNumeric MonthFactor  Year  RAV4Sales  Unemployment  RAV4Queries  \
0              1     January  2011      11196           9.1           29   
1              2    February  2011      12562           9.0           29   
2              3       March  2011      16082           9.0           29   
3              4       April  2011      15586           9.1           27   
4              5         May  2011       8624           9.0           28   
..           ...         ...   ...        ...           ...          ...   
67             8      August  2016      33171           4.9           59   
68             9   September  2016      29438           5.0           54   
69            10     October  2016      26429           4.9           52   
70            11    November  2016      28116           4.7           48   
71            12    December  2016      37214           4.7           54   

     CPIAll  CPIEnergy  LoanRate  
0   221.187    229.258       NaN  
1   221.898    232.068       NaN  
2   223.046    240.079       NaN  
3   224.093    247.977       NaN  
4   224.806    250.744       NaN  
..      ...        ...       ...  
67  240.595    189.718      4.18  
68  241.068    192.158      4.23  
69  241.641    195.541      4.26  
70  241.993    195.927      4.27  
71  242.712    200.139      4.32  

[72 rows x 9 columns]

In [120]:
model2 = smf.ols(formula='RAV4Sales ~ MonthFactor + Unemployment + RAV4Queries + LoanRate', data=data_train).fit()
print(model2.summary())

# cols = ['Unemployment', 'RAV4Queries']
# VIF(data_train, cols)

                            OLS Regression Results                            
Dep. Variable:              RAV4Sales   R-squared:                       0.844
Model:                            OLS   Adj. R-squared:                  0.739
Method:                 Least Squares   F-statistic:                     8.088
Date:                Tue, 09 Feb 2021   Prob (F-statistic):           1.38e-05
Time:                        18:43:24   Log-Likelihood:                -324.33
No. Observations:                  36   AIC:                             678.7
Df Residuals:                      21   BIC:                             702.4
Df Model:                          14                                         
Covariance Type:            nonrobust                                         
                               coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------------
Intercept               